bikestreets-ios/Managers/InstructionGenerator.swift

```diff
@@ -0,0 +1,343 @@
+      } else if currentStepDistance > advanceStepThreshold {
+        advanceDistanceAlongGeometry = currentStepDistance - instructionSpeakingDistance
+      }
+      if advanceDistanceAlongGeometry > 0 {
+        addVoiceInstruction(announcement: constructAdvanceAnnouncement(instructions: nextStepInfo?.stepInstructions, distanceString: DistanceConverter.convertToEnglishPhrase(from: advanceDistanceAlongGeometry + instructionSpeakingDistance)), distanceAlongGeometry: advanceDistanceAlongGeometry)
+      }
+      
+      // primary instruction
+      let nextNextStepInstructions = (nextNextStepManeuverType == arrival) ? lookAheadArrivalInstruction : nextNextStepInfo?.stepInstructions
+      let primaryCueAheadDistance: Double = min(currentStepDistance, (nextStepManeuverType == arrival) ? maxCueDistanceForArrival: maxCueDistanceForPrimaryInstruction)
+      
+      if let nextStepInstructions = nextStepInfo?.stepInstructions {
+        if let nextStepDistance, nextStepDistance < shortStepThreshold {
+          addVoiceInstruction(announcement: constructLookAheadAnnouncement(firstInstructions: nextStepInfo?.stepInstructions, nextInstructions: nextNextStepInstructions), distanceAlongGeometry: primaryCueAheadDistance)
+        } else {
+          addVoiceInstruction(announcement: nextStepInstructions, distanceAlongGeometry: primaryCueAheadDistance)
+        }
+      }
+      
+      return voiceInstructions
+    }
+  }
+  
+  
+  private static func constructDepartAnnouncement(instructions: String?, distanceString: String) -> String {
+    var departAnnouncement = ""
+    if let instructions = instructions {
+      departAnnouncement += "\(instructions) for \(distanceString)"
+    }
+    return departAnnouncement
+  }
+  
+  private static func constructContinueAnnouncement(currentStepName: String?, distanceString: String) -> String {
+    var continueAnnouncement = "Continue"
+    if let currentStepName = currentStepName {
+      continueAnnouncement += " on \(currentStepName) for \(distanceString)"
+    }
+    return continueAnnouncement
+  }
+  
+  private static func constructAdvanceAnnouncement(instructions: String?, distanceString: String) -> String {
+    var announcement = ""
+    if let instructions {
+      announcement = "In \(distanceString), \(instructions)"
+    }
+    return announcement
+  }
+  
+  private static func constructLookAheadAnnouncement(firstInstructions: String?, nextInstructions: String?) -> String {
+    var announcement = ""
+    if let firstInstructions {
+      announcement = firstInstructions
+      if let nextInstructions {
+        announcement += ", then \(nextInstructions)"
+      }
+    }
+    return announcement
+  }
+  
+}
+
+enum DistanceConverter {
+    static func convertToEnglishPhrase(from meters: Double) -> String {
+        let metersInAMile = 1609.34
+        let metersInAFeet = 0.3048
+
+        let miles = meters / metersInAMile
+        let feet = meters / metersInAFeet
```

Comment:
Not a big deal but we can probably use the Apple SDK here:
```suggestion
        let metersMeasurement = Measurement(value: meters, unit: UnitLength.meters)
        
        let miles = metersMeasurement.converted(to: UnitLength.miles)
        let feet = metersMeasurement.converted(to: UnitLength.feet)
```

Reply:
Agreed.

Reply:
I need to use .value at the end of your suggestion (so that I have the values as a Double), so I'll just include that modification to your suggestion in my commit.